onnxscript/ir/_protocols.py

```diff
@@ -254,9 +283,79 @@ class GraphProtocol(Protocol):
     doc_string: str
     opset_imports: Mapping[str, int]
     metadata_props: Mapping[str, str]
+    meta: Mapping[str, Any]
+
+    def __getitem__(self, index: int) -> NodeProtocol: ...
+    def __len__(self) -> int: ...
+    def __iter__(self) -> Iterator[NodeProtocol]: ...
+    def __reversed__(self) -> Iterator[NodeProtocol]: ...
+
+
+@typing.runtime_checkable
+class MutableGraphProtocol(Protocol):
+    """Topologically mutable graphs.
+
+    Graph represents a computation graph. In addition to the ONNX specification
+    specified fields, it also contains a mapping of :attr:`opset_imports`. This
+    allows different subgraphs to import different opsets. It is the responsibility
+    of the deserializer to reconcile the different opsets.
+
+    The :attr:`nodes` are not guaranteed to be topologically sorted. But the
+    iteration order should be deterministic across different runs. It is the
+    responsibility of the user to maintain a topological order of the nodes.
+
+    Attributes:
+        name: The name of the graph.
+        inputs: The input values of the graph.
+        outputs: The output values of the graph.
+        nodes: All nodes this graph directly owns. They do not have to be sorted.
+        initializers: The initializers in the graph.
+        doc_string: Documentation string.
+        opset_imports: Opsets imported by the graph.
+        metadata_props: Metadata.
+    """
+
+    # Block: Sync with NodeProtocol
+    # TODO(justinchuby): Support quantization_annotation
+    name: str | None
+    inputs: Sequence[ValueProtocol]
+    outputs: Sequence[ValueProtocol]
+    nodes: Sequence[NodeProtocol]
+    initializers: Mapping[str, TensorProtocol]
+    doc_string: str
+    opset_imports: Mapping[str, int]
+    metadata_props: Mapping[str, str]
+    meta: Mapping[str, Any]
 
-    def topologically_sorted_nodes(self) -> Sequence[NodeProtocol]:
-        """Return the nodes in topological order."""
+    def __getitem__(self, index: int) -> NodeProtocol: ...
+    def __len__(self) -> int: ...
+    def __iter__(self) -> Iterator[NodeProtocol]: ...
+    def __reversed__(self) -> Iterator[NodeProtocol]: ...
+
+    # Mutation methods
+    # End Block
+    def append(self, node: NodeProtocol) -> None:
+        """Append a node to the graph."""
+        ...
+
+    def extend(self, nodes: Iterable[NodeProtocol]) -> None:
+        """Extend the graph with the given nodes."""
+        ...
+
+    def remove(self, node: NodeProtocol) -> None:
+        """Remove a node from the graph."""
+        ...
+
+    def insert_after(self, node: NodeProtocol, new_nodes: Iterator[NodeProtocol]) -> None:
+        """Insert new nodes after the given node."""
+        ...
+
+    def insert_before(self, node: NodeProtocol, new_nodes: Iterator[NodeProtocol]) -> None:
+        """Insert new nodes before the given node."""
+        ...
+
+    def sort(self) -> None:
```

Comment:
The choices are offering as a core api like this, or as a convenience function in a separate module. I think it is reasonable to assume that when users call sort() they are conscious about the runtime implication. It is also an obvious line to be spotted and optimized away.

Reply:
I recall we settled on requesting users to maintain topological sorted nodes when writing passes. IMO it is a trade-off between ease of use and performance. I'm still not sure it is worth the extra perf overhead and complexity brought in with sort.

Reply:
Right - the users are required to maintain the order. They can do it in two ways: (1) maintain the order at all times or (2) insert anywhere then sort. It would be the user's choice on how they want to maintain the order, but we can provide a sort() users can call.

Reply:
Option (2) is not a practical option if we do not provide sort api. Users won't do it if they have to implement sort on our custom linked list data structure themselves.
Both my comments are around concerns of adding sort api that makes option (2) feasible, and whether the trade-off is worth it.

Reply:
Sounds good to me. @BowenBao : would you prefer topological sort be an external utility, not a class method? I am not sure I understand the concern. It is useful to have a topological sort utility somewhere ... ORT's graph has it; onnx users ask for it for onnx models (even though the spec says nodes must be sorted, not all tools produce sorted nodes). It is not being called implicitly, only explicitly by an user.

Reply:
@gramalingam I'm hesitant for this utility to exist. Do we know use cases where sort is practically required, and not just making things more convenient?

Reply:
Yeah for now it can stay unimplemented if we prefer. IMO users are going to need the sorting capability. Sometimes it is easier and potentially even cheaper to just sort at the end. I would prefer implementing a correct and tested sort then to let users create multiple versions of sorts on their own which can be not as well tested.

That said, if a user need to implement sort they will not need to understand the linked list. They can simply sort the nodes and call `extend()` and supply the sorted list.